mypy/partially_defined.py

```diff
@@ -597,7 +603,7 @@ def visit_starred_pattern(self, o: StarredPattern) -> None:
         super().visit_starred_pattern(o)
 
     def visit_name_expr(self, o: NameExpr) -> None:
-        if refers_to_builtin(o):
+        if o.name in self.builtins:
```

Comment:
Hmm this should only affect references to global definitions. Can you add a test case to ensure that a local variable named `type` is still processed? A potential fix would be to check if `fullname` starts with `builtins.`.

Reply:
It think that you can also check that the `kind` of `SymbolTableNode` is `GDEF`.